File: app/src/main/java/org/fossasia/susi/ai/activities/ForgotPasswordActivity.java
package org.fossasia.susi.ai.activities;
import android.app.ProgressDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.Color;
import android.os.Bundle;
import android.support.design.widget.TextInputLayout;
import android.support.v7.app.AppCompatActivity;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.RadioButton;

import org.fossasia.susi.ai.R;
import org.fossasia.susi.ai.helper.Constant;
import org.fossasia.susi.ai.helper.AlertboxHelper;
import org.fossasia.susi.ai.helper.CredentialHelper;
import org.fossasia.susi.ai.helper.PrefManager;
import org.fossasia.susi.ai.rest.ClientBuilder;
import org.fossasia.susi.ai.rest.responses.susi.ForgotPasswordResponse;

import java.net.UnknownHostException;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

/**
 * <h1>The Forgot password activity.</h1>
 * <h2>This activity is used in case user forgets his/her password and wants to recover it.</h2>
 */
public class ForgotPasswordActivity extends AppCompatActivity {

    @BindView(R.id.forgot_email)
    protected TextInputLayout email;
    @BindView(R.id.reset_button)
    protected Button resetButton;
    @BindView(R.id.susi_default)
    protected RadioButton susiServer;
    @BindView(R.id.personal_server)
    protected RadioButton personalServer;
    @BindView(R.id.input_url)
    protected TextInputLayout url;

    private String alertTitle,alertMessage;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_forgot_password);
        ButterKnife.bind(this);
        setTitle(getString(R.string.forgot_pass_activity));

        if (savedInstanceState != null) {
            email.getEditText().setText(savedInstanceState.getCharSequenceArray(Constant.SAVED_STATES)[0].toString());
            if(savedInstanceState.getBoolean(Constant.SERVER)) {
                url.setVisibility(View.VISIBLE);
            } else {
                url.setVisibility(View.GONE);
            }
        }

        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            finish();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    /**
     * Show input edit text for custom url.
     */
    @OnClick(R.id.personal_server)
    public void showURL() {
        url.setVisibility(View.VISIBLE);
    }

    /**
     * Hide input edit text for custom url.
     */
    @OnClick(R.id.susi_default)
    public void hideURL() {
        url.setVisibility(View.GONE);
    }

    /**
     * Called when a user clicks on the reset up button
     */
    @OnClick(R.id.reset_button)
    public void signUp() {
        if (CredentialHelper.checkIfEmpty(email, this)) {
            return;
        }

        if (!CredentialHelper.isEmailValid(email.getEditText().getText().toString())) {
            email.setError(getString(R.string.email_invalid_title));
            return;
        }

        if(personalServer.isChecked()) {
            if(!CredentialHelper.checkIfEmpty(url,this) && CredentialHelper.isURLValid(url,this)) {
                if (CredentialHelper.getValidURL(url,this) != null) {
                    PrefManager.putBoolean(Constant.SUSI_SERVER, false);
                    PrefManager.putString(Constant.CUSTOM_SERVER, CredentialHelper.getValidURL(url, this));
                } else {
                    url.setError(getApplicationContext().getString(R.string.invalid_url));
                    return;
                }
            } else {
                return;
            }
        } else{
            PrefManager.putBoolean(Constant.SUSI_SERVER, true);
        }

        email.setError(null);
        resetButton.setEnabled(false);
        final ProgressDialog progressDialog = new ProgressDialog(this);
        progressDialog.setCancelable(false);
        progressDialog.setMessage(getApplicationContext().getString(R.string.loading));
        progressDialog.show();

        final Call<ForgotPasswordResponse> forgotPasswordResponseCall = new ClientBuilder().getSusiApi()
                .forgotPassword(email.getEditText().getText().toString().trim());
        progressDialog.setOnCancelListener(new DialogInterface.OnCancelListener() {
            @Override
            public void onCancel(DialogInterface dialog) {
                forgotPasswordResponseCall.cancel();
                resetButton.setEnabled(true);
            }
        });

        forgotPasswordResponseCall.enqueue(new Callback<ForgotPasswordResponse>() {
            @Override
            public void onResponse(Call<ForgotPasswordResponse> call, Response<ForgotPasswordResponse> response) {
                if (response.isSuccessful() && response.body() != null) {
                    alertSuccess(response);
                } else if(response.code() == 422) {
                    alertTitle = getResources().getString(R.string.email_invalid_title);
                    alertMessage = getResources().getString(R.string.email_invalid);
                    alertNotSuccess(alertTitle, alertMessage, getResources().getString(R.string.retry), Color.RED);
                } else {
                    alertNotSuccess(response.code()+getResources().getString(R.string.error), response.message(), getResources().getString(R.string.ok), Color.BLUE);
                }
                resetButton.setEnabled(true);
                progressDialog.dismiss();
            }

            @Override
            public void onFailure(Call<ForgotPasswordResponse> call, Throwable t) {
                t.printStackTrace();
                if(t instanceof UnknownHostException) {
                    alertTitle = getResources().getString(R.string.unknown_host_exception);
                    alertMessage = t.getMessage();
                } else {
                    alertTitle = getResources().getString(R.string.error_internet_connectivity);
                    alertMessage = getResources().getString(R.string.no_internet_connection);
                }
                alertNotSuccess(alertTitle,alertMessage, getResources().getString(R.string.retry), Color.RED);
                resetButton.setEnabled(true);
                progressDialog.dismiss();
            }
        });
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        CharSequence values[] = {email.getEditText().getText().toString()};
        outState.putCharSequenceArray(Constant.SAVED_STATES, values);
        outState.putBoolean(Constant.SERVER,personalServer.isChecked());
    }

    /**
     * Displays an alert dialog box in case of successful reset of password.
     *
     * @param response the response from server
     */
    public void alertSuccess(Response<ForgotPasswordResponse> response) {
        DialogInterface.OnClickListener dialogClickListener = new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int id) {
                Intent intent = new Intent(ForgotPasswordActivity.this, LoginActivity.class);
                intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_NEW_TASK);
                startActivity(intent);
                finish();
            }
        };
        alertTitle = getResources().getString(R.string.forgot_password_mail_sent);
        alertMessage = response.body().getMessage();
        AlertboxHelper successAlertboxHelper = new AlertboxHelper(ForgotPasswordActivity.this, alertTitle, alertMessage, dialogClickListener, null, getResources().getString(R.string.Continue), null, Color.BLUE);
        successAlertboxHelper.showAlertBox();
    }

    /**
     * Displays an alert dialog box in case of unsuccessful reset of password.
     *
     * @param title              the title of dialog box
     * @param message            the body of dialog box
     * @param positiveButtonText the positive button text
     * @param colour             the colour of buttons
     */
    public void alertNotSuccess(String title,String message,String positiveButtonText,int colour) {
        AlertboxHelper notSuccessAlertboxHelper = new AlertboxHelper(ForgotPasswordActivity.this, title, message, null, null, positiveButtonText, null, colour);
        notSuccessAlertboxHelper.showAlertBox();
    }
}

File: app/src/main/java/org/fossasia/susi/ai/forgotpassword/ForgotPasswordActivity.kt
package org.fossasia.susi.ai.forgotpassword

import android.app.ProgressDialog
import android.content.DialogInterface
import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.MenuItem
import android.view.View
import kotlinx.android.synthetic.main.activity_forgot_password.*
import org.fossasia.susi.ai.R
import org.fossasia.susi.ai.forgotpassword.contract.IForgotPasswordPresenter
import org.fossasia.susi.ai.forgotpassword.contract.IForgotPasswordView
import org.fossasia.susi.ai.helper.AlertboxHelper
import org.fossasia.susi.ai.helper.Constant
import org.fossasia.susi.ai.login.LoginActivity

/**
 * <h1>The ForgotPassword activity.</h1>
 * <h2>This activity is used to request password if user forgot password.</h2>
 *
 * Created by meeera on 6/7/17.
 */
class ForgotPasswordActivity : AppCompatActivity(), IForgotPasswordView {

    lateinit var forgotPasswordPresenter: IForgotPasswordPresenter
    lateinit var progressDialog: ProgressDialog

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_forgot_password)
        title = getString(R.string.forgot_pass_activity)

        if (savedInstanceState != null) {
            forgot_email.editText?.setText(savedInstanceState.getCharSequenceArray(Constant.SAVED_STATES)[0].toString())
            if (savedInstanceState.getBoolean(Constant.SERVER)) {
                input_url.visibility = View.VISIBLE
            } else {
                input_url.visibility = View.GONE
            }
        }

        val bundle = intent.extras;
        val string = bundle?.getString("email")
        if (string != null)
            forgot_email.editText?.setText(string)

        if (supportActionBar != null) {
            supportActionBar?.setDisplayHomeAsUpEnabled(true)
        }

        progressDialog = ProgressDialog(this)
        progressDialog.setCancelable(false)
        progressDialog.setMessage(getString(R.string.reset_password_message))

        showUrl()
        cancelRequestPassword()
        requestPassword()

        forgotPasswordPresenter = ForgotPasswordPresenter(this)
        forgotPasswordPresenter.onAttach(this)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finish()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    override fun success(title: String?, message: String?) {
        val dialogClickListener = DialogInterface.OnClickListener { _, _ ->
            val intent = Intent(this@ForgotPasswordActivity, LoginActivity::class.java)
            intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_NEW_TASK
            startActivity(intent)
            finish()
        }
        val successAlertboxHelper = AlertboxHelper(this@ForgotPasswordActivity, title, message, dialogClickListener, null, resources.getString(R.string.Continue), null, Color.BLUE)
        successAlertboxHelper.showAlertBox()
        reset_button.isEnabled = true
    }

    override fun failure(title: String?, message: String?, button: String?, color: Int) {
        val notSuccessAlertboxHelper = AlertboxHelper(this@ForgotPasswordActivity, title, message, null, null, button, null, color)
        notSuccessAlertboxHelper.showAlertBox()
        reset_button.isEnabled = true
    }

    fun showUrl() {
        custom_server.setOnClickListener {
            input_url.visibility = if (custom_server.isChecked) View.VISIBLE else View.GONE
        }
    }

    fun cancelRequestPassword() {
        progressDialog.setOnCancelListener {
            forgotPasswordPresenter.cancelSignup()
            reset_button.isEnabled = true
        }
    }

    fun requestPassword() {
        reset_button.setOnClickListener {
            val email = forgot_email.editText?.text.toString()
            val isPersonalServerChecked = custom_server.isChecked
            val url = input_url.editText?.text.toString()
            forgot_email.error = null
            input_url.error = null
            reset_button.isEnabled = false
            forgotPasswordPresenter.requestPassword(email, url, isPersonalServerChecked)
        }
    }

    override fun invalidCredentials(isEmpty: Boolean, what: String) {
        if (isEmpty) {
            when (what) {
                Constant.EMAIL -> forgot_email.error = getString(R.string.email_cannot_be_empty)
                Constant.INPUT_URL -> input_url.error = getString(R.string.url_cannot_be_empty)
            }
        } else {
            when (what) {
                Constant.EMAIL -> forgot_email.error = getString(R.string.email_invalid_title)
                Constant.INPUT_URL -> input_url.error = getString(R.string.invalid_url)
            }
        }
        reset_button.isEnabled = true
    }

    override fun showProgress(boolean: Boolean) {
        if (boolean) progressDialog.show() else progressDialog.hide()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        val values = arrayOf<CharSequence>(forgot_email.editText?.text.toString())
        outState.putCharSequenceArray(Constant.SAVED_STATES, values)
        outState.putBoolean(Constant.SERVER, custom_server.isChecked)
    }

    override fun onDestroy() {
        super.onDestroy()
        forgotPasswordPresenter.onDetach()
    }
}